File: app/src/main/java/eu/darken/bluetoothmanager/core/BluetoothEventReceiver.java
package eu.darken.bluetoothmanager.core;

import android.bluetooth.BluetoothDevice;
import android.content.Context;
import android.content.Intent;
import android.support.v4.content.WakefulBroadcastReceiver;

import timber.log.Timber;


public class BluetoothEventReceiver extends WakefulBroadcastReceiver {
    public static final String EXTRA_ACTION = "eu.darken.ffs.spotify.bluetooth.action";
    public static final String EXTRA_DEVICE = "eu.darken.ffs.spotify.bluetooth.device";

    @Override
    public void onReceive(Context context, Intent intent) {
        Timber.d("onReceive(%s,%s)", context, intent);
        Timber.d("Device: %s | Action: %s", intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE), intent.getAction());
        Intent service = new Intent(context, EventService.class);
        service.putExtra(EXTRA_ACTION, intent.getAction());
        service.putExtra(EXTRA_DEVICE, (BluetoothDevice) intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE));
        startWakefulService(context, service);
    }
}

File: app/src/main/java/eu/darken/bluemusic/monitor/core/receiver/BluetoothEventReceiver.kt
package eu.darken.bluemusic.monitor.core.receiver

import android.bluetooth.BluetoothDevice
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import dagger.hilt.android.AndroidEntryPoint
import eu.darken.bluemusic.common.coroutine.AppScope
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.WARN
import eu.darken.bluemusic.common.debug.logging.log
import eu.darken.bluemusic.common.debug.logging.logTag
import eu.darken.bluemusic.monitor.core.worker.MonitorControl
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class BluetoothEventReceiver : BroadcastReceiver() {

    @Inject lateinit var monitorControl: MonitorControl
    @Inject @AppScope lateinit var appScope: CoroutineScope

    override fun onReceive(context: Context, intent: Intent) {
        log(TAG) { "onReceive($context, $intent)" }
        if (!EXPECTED_ACTIONS.contains(intent.action)) {
            log(TAG, WARN) { "Unknown action: ${intent.action}" }
            return
        }

        val bluetoothDevice = intent.getParcelableExtra<BluetoothDevice>(BluetoothDevice.EXTRA_DEVICE)
        if (bluetoothDevice == null) {
            log(TAG, WARN) { "Event without Bluetooth device association." }
            return
        } else {
            log { "Event related to $bluetoothDevice" }
        }

        val pending = goAsync()
        appScope.launch {
            log(TAG) { "Starting monitor" }
            monitorControl.startMonitor(forceStart = false)
            pending.finish()
        }
    }

    companion object {
        private val TAG = logTag("Monitor", "EventReceiver")
        private val EXPECTED_ACTIONS = setOf(
            BluetoothDevice.ACTION_ACL_CONNECTED,
            BluetoothDevice.ACTION_ACL_DISCONNECTED,
        )
    }
}
